_posts/2018-02-06-unrealiable-network.md

```diff
@@ -0,0 +1,116 @@
+                            (if (some? response)
+                              ;; Successful response, return it to the caller by using
+                              ;; the given response channel.
+                              (do (put! response-chan response)
+                                  (close! response-chan))
+                              ;; Failing response, any attempts left?
+                              (if (> (:attempts re-try-options) 1)
+                                ;; Try again, decrease attempts by one and increase timeout
+                                (network-request
+                                  response-chan
+                                  request-options
+                                  {:timeout (+ (:timeout re-try-options) 2000)
+                                   :attempts (dec (:attempts re-try-options))})
+                                ;; Return the failing response to the caller
+                                (do (put! response-chan response)
+                                    (close! response-chan)))))]
+     ;; Send the network request, return the response to the response-handler
+     (go
+       ;; Before sending, wait for the timeout (if given)
+       (when-let [re-try-timeout (:timeout re-try-options)]
+         (<! (timeout re-try-timeout)))
+       (ajax-request (merge
+                       request-options
+                       {:handler response-handler}))))))
+```
+
+In most cases, however, it is not practical to re-try every single failed request. For exampole, if a request pointed to a file which was not found and returned 404, it probably does not help to to try this request again. Thus, you should choose which requests you might want to try again.
+
+# Cache Responses When Necessary
+
+Caching is our friend when we want to make things work faster, but it is also a good way to help surviving with unreliable network connections. Our ultimate goal here is to avoid making requests that would probably return the same result as we got before. Caching can be used both client and server side. In most cases, we want to trust the browser and the server to handle caching by using proper [HTTP caching headers](https://developer.mozilla.org/en-US/docs/Web/HTTP/Caching). Some cases, though, need special attention and application level optimisation.
+
+On the client side, you could ask yourself if making a specific request with the same parameters multiple times on short intervals is absolutely necessary. If the server is probably going to return the same answer, you might consider caching the request's timestamp, parameters and response. If the same request is made again, with the same parameters, shortly after the previous one, you can simply return the answer from the cache, without making network requests. The downside is, of course, that the user does not get the most recent data from the server, so you should use clientside caching only when this is not a problem.
```

Comment:
> you could ask yourself if making

you could ask yourself whether making

> If the server is probably going to return the same answer

If the server is likely going to return the same answer

> clientside

client-side